subsys/jwt/jwt.c

```diff
@@ -317,3 +395,125 @@ int jwt_init_builder(struct jwt_builder *builder,
 
 	return jwt_add_header(builder);
 }
+
+int jwt_parse_payload(struct jwt_parser *parser, int32_t *expt, int32_t *iat, char *aud)
+{
+	struct jwt_payload payload;
+	int res;
+
+	res = base64url_decode(parser->buf, parser->buf_len, parser->payload, parser->payload_len);
+	if (res < 0) {
+		return res;
+	}
+
+	res = json_obj_parse(parser->buf, res, jwt_payload_desc, ARRAY_SIZE(jwt_payload_desc),
+			     &payload);
+	if (res == (1 << ARRAY_SIZE(jwt_payload_desc)) - 1) {
+		*expt = payload.exp;
+		*iat = payload.iat;
+		strcpy(aud, payload.aud);
+		res = 0;
+	} else if (res >= 0) {
+		res = -EINVAL;
+	}
+
+	return res;
+}
+
+int jwt_verify(struct jwt_parser *parser, const char *der_key, size_t der_key_len)
+{
+	struct jwt_builder builder;
+	char *builder_header;
+	size_t builder_header_len;
+	char *builder_sign;
+	size_t builder_sign_len;
+	int res;
+
+	/*
+	 * Use existing jwt_add_header and jwt_sign functions to fill a
+	 * jwt_builder structure and use it to compare header and signature
+	 * fields
+	 */
+
+	/*
+	 * Check JWT header
+	 */
+
+	res = jwt_init_builder(&builder, parser->buf, parser->buf_len);
+	if (res != 0) {
+		return res;
+	}
+	builder_header = builder.base;
+	builder_header_len = builder.buf - builder.base;
+	res = strncmp(parser->header, builder_header, builder_header_len);
+	if (res != 0) {
+		return -EINVAL;
+	}
+
+	/*
+	 * Copy the payload
+	 */
+
+	if (builder.len < (parser->payload_len + 1)) {
+		return -ENOSPC;
+	}
+	/* -1/+1 to also copy the dot */
+	strncpy(builder.buf, parser->payload - 1, parser->payload_len + 1);
```

Comment:
Likewise here, `strncpy` is definitely not the right thing to use, as it can leave the buffer not terminated. In this particular case, it _always_ leave the buffer non-terminated. I'd just use `memcpy` of the given length, and place the terminating nul manually.

Reply:
OK, changed